### PR TITLE
3692 incremental builds, joplin side

### PR DESCRIPTION
**Pull request by chiaberry**
# Description

Hello!

The meat of the work on this issue can be found in `janis_build_triggers`. When a page is published or unpublished, there is a hook that triggers a build. Thanks to Eric's note, I added a flag so we can toggle collecting the pages, turning on incremental builds. 

Using get_object_usage, we check to see where that page is being referenced, and then collecting all of those page ids to eventually send along to Publisher. Through my work on this, I realized that not all of our pages know their relationships, primarily Service and Information Pages don't know they are on guides. I have a function `find_pages_in_guides` that goes through all of our guide pages and checks to see if the page we changed is in one of the sections. 

Updating snippets follow a similar pattern to the updated pages, with just some slight differences on functions. 

I included a test file testing some of our scenarios. Some of the tests currently fail, such as updating a map snippet or updating a Topic Page. Pages that are references on contextual nav don't know that they are, or who their siblings are. Use ` pipenv run joplin/manage.py test base.tests.test_incremental_builds --keepdb` to run the test. 

Another way to test this branch is to run undockered and put breakpoints/print in janis build triggers to check that your page id is coming back as expected, by then looking it up in localhost:8000/api/graphiql. 

In my journey through Joplin, I took the liberty to remove all imports and references to things we no longer use in our models. 

I tried to remove the entire ProcessPage model from `base.models.__init__` and translations, but got some errors I believe based on migrations. I left a note. 

# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How can this be tested?

 

# Checklist:
- [x] Request reviewers
- [x] Slack-out message for review
- [x] Link this PR in the issue
- [x] Moved card to *review* in zenhub

### CODE REVIEW

File: joplin/base/signals/janis_build_triggers.py [hunk 1]
```diff
@@ -19,11 +17,12 @@
 JANIS_SLUG_URL = settings.JANIS_SLUG_URL
 
 
-def trigger_build(sender, action='saved', instance=None):
+def trigger_build(sender, pages_ids, action='saved', instance=None):
```
Comment (easherma) [inline joplin/base/signals/janis_build_triggers.py]: We should probably set this up on a feature flag, so we retain the ability to _not_ publish incrementally, plus that will let us merge the code into staging sooner

File: joplin/api/schema.py [hunk 2]
```diff
@@ -135,11 +132,11 @@ class Meta:
     filename = graphene.String()
 
 
-class ThreeOneOneNode(DjangoObjectType):
-    class Meta:
-        model = ThreeOneOne
-        filter_fields = ['title']
-        interfaces = [graphene.Node]
+# class ThreeOneOneNode(DjangoObjectType):
```
Comment (briaguya) [inline joplin/api/schema.py]: let's go ahead and just remove this completely!

File: joplin/base/models/__init__.py [hunk 3]
```diff
@@ -68,6 +60,7 @@ def __str__(self):
         return self.title
 
 
+# I tried to remove the ProcessPages, but ended up getting errors with old migrations. - chia 1/22/20
```
Comment (briaguya) [inline joplin/base/models/__init__.py]: will this become a "comment with hours wasted trying to remove process pages" type thing? the world may never know...

File: joplin/base/models/topic_page.py [hunk 4]
```diff
@@ -52,8 +50,9 @@ class TopicPageTopPage(Orderable):
         PageChooserPanel('page', page_type=[InformationPage, ServicePage, GuidePage, OfficialDocumentPage]),
     ]
 
-    def __str__(self):
-        return self.page.text
+    # this errors because it says page has no attribute text, can we fall back on the original method?
+    # def __str__(self):
+    #     return self.page.text
```
Comment (briaguya) [inline joplin/base/models/topic_page.py]: maybe we can just use `self.page.title`?

File: joplin/base/signals/janis_build_triggers.py [hunk 5]
```diff
@@ -32,6 +31,65 @@ def trigger_build(sender, action='saved', instance=None):
...
+    :return: id of the Guide Pages that include that page
+    """
+    pages_id = []
+    all_guides = GuidePage.objects.all()
+    for g in all_guides:
+        for s in g.sections:
+            # s.value.items() is an ordered dict
+            list_of_values = list(s.value.items())
+            # the pages are the 5th tuple
+            pages = list_of_values[4][1]
```
Comment (briaguya) [inline joplin/base/signals/janis_build_triggers.py]: I'm assuming you've tried this already but just confirming that list_of_values["key_name"] doesn't work

**Pull request merged**